Add event_attribution_reporting.md to support views (#129)
WICG/conversion-measurement-api: event_attribution_reporting.md

### Attribution Triggering

Attribution trigger redirects will be able to specify different data to be used, depending on the `source_type` of any associated sources.

This is done via a new query param, `event-source-trigger-data` on the trigger [redirect](https://github.com/WICG/conversion-measurement-api#triggering-attribution). This data will be limited to 1 bit. This data will be noised 5% of the time, matching the [noise](https://github.com/WICG/conversion-measurement-api#data-limits-and-noise) for navigation sources.

Event sources will only be able to generate a single attribution report, compared to the three reports for navigation sources. Alongside the 1 bit of data, this further restricts the total amount of trigger side data which can be associated with a single `attributionsourceeventid`. See [privacy considerations](#privacy-considerations) for why this is necessary.

### Coarse Attribution Expiry

Event sources will have their `attributionexpiry` rounded to the nearest whole day.

> The `attributionexpiry` of a source can be used to associate attributionsourceeventids with a coarse trigger side timestamp. Because event sources are easier to register en masse than navigation sources, it would allow a reporting origin to do this with a finer granularity.

Reviewer: ```suggestion
The `attributionexpiry` of a source can be used to associate `attributionsourceeventid`s with a coarse trigger side timestamp. Because event sources are easier to register en masse than navigation sources, it would allow a reporting origin to do this with a finer granularity.
```

These steps will not be performed for navigation attribution sources.

See the [privacy considerations section](#privacy-considerations) for the rationale behind adding this noise.

### Controlling which attribution source to trigger

In the Click Through Explainer, at triggering time, the browser looks up all matching sources in storage. A [credit](https://github.com/WICG/conversion-measurement-api#multiple-sources-for-the-same-trigger-multi-touch) of 100 is assigned to the most recent matching source in storage, and a credit of 0 is assigned to the rest. Reports are sent for each of these. We propose changing this for both click sources and event sources.

To provide additional utility in prioritizing the distribution of credit between different classes of sources, the browser can associate a priority with each attribution source. This priority will be defined in a new attribute on attribution source tags, `attributionsourcepriority`, which holds a 64 bit integer.

When a trigger redirect is received, the browser will find the matching source with highest `attributionsourcepriority` value and generate a report. The other sources will not generate reports.

> If reports were sent for these sources, it would be possible for `reportingorigin` to use these other reports to remove the noise on the attribution triggering event. 

Reviewer: ```suggestion
If reports were sent for these sources, it would be possible for `reportingorigin` to use these other reports to remove the noise on the attribution triggering event, unless we introduced a mechanism to noise the sources dependently on others. 
```

Further, the addition of event noising protects against any cross site data leakage based on `p`.

### Less trigger side data

Registering event attribution sources is not gated on a user interaction or top level navigation, allowing them to be registered more frequently and with greater ease. For example, by restricting to 1 bit of data and 1 report per source, a `reportingorigin` would need to register many more sources in order to link cross-site identity relative to the Click Through API.

This is further restricted by rate limiting the usage of the API between two sites, using [reporting cooldowns](https://github.com/WICG/conversion-measurement-api#reporting-cooldown). Due to the different characteristics between classes of sources, these cooldowns should have independent limits on the number of reports of each type.

The number of reporting windows is another vector which can contain trigger side information. By restricting to a [single window](#single-reporting-window), a `reportingorigin` does not receive any additional information on when in the attribution window a source was triggered.

## Browsing history reconstruction

> Reporting attribution without a pre-existing navigation allows the `reportingorigin` to learn whether a given user on the source site visited the attributiondestination site at all. In the Click through API, this is not an issue because the `reportingorigin` knows a priori the user was navigating to attributiondestination.

Reviewer: ```suggestion
Reporting attribution without a pre-existing navigation allows the `reportingorigin` to learn whether a given user on the source site visited the `attributiondestination` site at all. For click-through reports, this is not an issue because the `reportingorigin` knows a priori the user was navigating to `attributiondestination`.
```

This new threat is/can be mitigated in a number of ways:

### Plausible Deniability

By adding noise to whether an attribute source gets [triggered](#noise-on-the-attribution-triggering-event), a `reportingorigin` will not know with absolute certainty whether a user actually visited a site. In fact, this additional mitigation gives some local differential privacy guarantees on the data in the API.

It becomes important for browsers to ensure that other API parameters do not allow a `reportingorigin` to easily denoise reports for a given `attributionsourceeventid`. For example, if multiple reports were sent for each source it may allow reporters to use the number of reports as a mechanism to remove noise, and may require browsers to implement more complex noise logic.

Note: Since conversion rates in practice are extremely rare events, we need to be careful to set the noise parameter so as not to ruin the utility of this API (i.e. noise drowns out signal).

### Limiting the number of unique sites covered by pending sources

> To limit the breadth of attributiondestination sites that a `reportingorigin` may be trying to measure user visits on, the browser can limit the number attributiondestination eTLD+1s represented by pending sources for a publisher.

Reviewer: ```suggestion
To limit the breadth of `attributiondestination` sites that a `reportingorigin` may be trying to measure user visits on, the browser can limit the number `attributiondestination` eTLD+1s represented by pending sources for a publisher.
```

> The browser can place a limit on the number of a source site's pending source's unique attributiondestinations. When an attribution source is registered for an eTLD+1 that is not already in the pending sources and a source site is at its limit, the browser will delete the oldest pending attribution source from the source site. If after deleting the oldest source, the source site is still at its limit, the browser will continue to delete pending sources until it is able to make room for the newly registered source.

Reviewer: ```suggestion
The browser can place a limit on the number of a source site's pending source's unique `attributiondestination`s. When an attribution source is registered for an eTLD+1 that is not already in the pending sources and a source site is at its limit, the browser will delete the oldest pending attribution source from the source site. If after deleting the oldest source, the source site is still at its limit, the browser will continue to delete pending sources until it is able to make room for the newly registered source.
```

> The lower this value, the harder it is for a `reportingorigin` to use the API to try and measure user browsing activity not associated with ads being shown. Browsers may choose parameters on the number of attributiondestinations to make their own tradeoffs for privacy and utility.

Reviewer: ```suggestion
The lower this value, the harder it is for a `reportingorigin` to use the API to try and measure user browsing activity not associated with ads being shown. Browsers may choose parameters on the number of `attributiondestination`s to make their own tradeoffs for privacy and utility.
```

> Because this limit is per source site, it is possible for different reportingorigins on a site to push the other attribution sources out of the browser. To limit abuse by potentially bad actors, the browser could consider adding a publisher side configuration which could partition this queue into smaller queues managed per reportingorigin.

Reviewer: ```suggestion
Because this limit is per source site, it is possible for different `reportingorigin`s on a site to push the other attribution sources out of the browser. To limit abuse by potentially bad actors, the browser could consider adding a source-site side configuration which could partition this queue into smaller queues managed per `reportingorigin`.
```

Reviewer: Was thinking possibly using 'source-site' instead of publisher but I don't have a strong preference either way.

Author: source-site seems better. Adjusted above as well.
